### PR TITLE
Batch index delete cluster state updates

### LINKED ISSUE #90022: Batch index delete cluster state tasks
### Description

Currently every index delete task is processed individually. Additionally, they are marked at the `urgent` level. This leads to scenarios where a number of index delete tasks that are queued up very quickly can lead to starvation. We have seen these tasks delaying the processing of `node-join` tasks disrupting cluster state processing.

### PR DESCRIPTION
**Pull request by original-brownbear**
Batching these up in a straight-forward manner using the new infrastructure for batched CS updates.

Mainly just me wanting to try and use the new batching API once to see if I get it right :)

closes #90022

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
https://github.com/elastic/elasticsearch/issues/89192 should cut out some of this boilerplate; I left a couple of other comments too.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java [hunk 1]
```diff
@@ -48,34 +45,37 @@ public class MetadataDeleteIndexService {
...
+                    state = deleteIndices(state, Sets.newHashSet(taskContext.getTask().indices()));
+                    taskContext.success(taskContext.getTask());
+                } catch (Exception e) {
+                    taskContext.onFailure(e);
+                }
+            }
+            if (state == batchExecutionContext.initialState()) {
+                return state;
+            }
+            return allocationService.reroute(state, "deleted indices");
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:68]: I believe in other places such as AutoCreateAction we do:
```suggestion
                    try (var ignored = batchExecutionContext.dropHeadersContext()) {
                        return allocationService.reroute(state, "deleted indices");
                    }
```

File: server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java [hunk 2]
```diff
@@ -115,9 +114,6 @@ public void testDeleteUnassigned() {
         assertNull(after.metadata().getIndices().get(index));
         assertNull(after.routingTable().index(index));
         assertNull(after.blocks().indices().get(index));
-
-        // Make sure we actually attempted to reroute
-        verify(allocationService).reroute(any(ClusterState.class), any(String.class));
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java]: Ah thanks, reworked the test using that class now and brought back this assertion.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java]: Thanks, applied :)

**Comment by original-brownbear** [comment]
Thanks you two, applied your points and brought back the test now.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
Actually sorry missed a thing: this deserves an entry in the release notes I think so not a `>non-issue`. I'll relabel it which should trigger the changelog creation.

**Review by DaveCTurner — APPROVED** [review]
Ok all good now.

**Comment by original-brownbear** [comment]
Thanks both!

**Pull request merged**